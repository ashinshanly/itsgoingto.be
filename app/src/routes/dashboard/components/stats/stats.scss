@import "variables";

.info {
  display: -ms-flexbox;
  display: flex;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-right: -15px;
  margin-left: -15px;
  margin-bottom: 2em;

  .info-block-column {
    position: relative;
    min-height: 1px;
    padding-right: 15px;
    padding-left: 40px;
    -ms-flex: 0 0 25%;
    flex: 0 0 25%;
    max-width: 25%;
  }

  .info-block-column:before {
    content: "";
    position: absolute;
    left: 0;
    height: 100%;
    border-left: 2px solid lighten($base-font-color, 30%);
  }

  .info-block-column:first-child:before {
    border-left: 0px;
  }

  .info-body {
    font-weight: 600;
    font-size: 3em;
    line-height: 1em;
  }
}
