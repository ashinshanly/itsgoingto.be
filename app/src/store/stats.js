import { prop, compose } from 'ramda'

// ------------------------------------
// Constants
// ------------------------------------
export const STATS_UPDATE = 'STATS_UPDATE'

// ------------------------------------
// Selectors
// ------------------------------------
/**
 * Returns a stat from the state
 *
 * @param  {object} state  App state
 * @param  {object} string The stat to select
 *
 * @return {boolean}       The stat
 */
export const statSelector = (state, stat) =>
  compose(
    prop(stat),
    prop('stats')
  )(state)

// ------------------------------------
// Actions
// ------------------------------------
/**
 * Update the stats in the state
 *
 * @param  {object} stats The user data
 *
 * @return {Function}     redux-thunk callable function
 */
export const updateStats = (stats) => ({
  type: STATS_UPDATE,
  stats,
})

/**
 * Clear the stats in the state
 *
 * @return {Function}    redux-thunk callable function
 */
export const clearStats = () => updateStats(initialState)

export const actions = {
  updateStats,
  clearStats,
}

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  // Insert the user into the state
  [STATS_UPDATE] : (previousState, action) => action.stats,
}

// ------------------------------------
// Reducer
// ------------------------------------
/**
 * Initial state for this store component
 */
const initialState = {
  polls:     null,
  responses: null,
}

/**
 * The reducer for this store component
 *
 * @param  {object} state  The current state
 * @param  {object} action The action to perform on the state
 *
 * @return {object}        The modified state
 */
export default function statsReducer (state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type]

  return handler ? handler(state, action) : state
}
