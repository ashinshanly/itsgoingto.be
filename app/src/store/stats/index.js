import { STATS_UPDATE } from './actions'

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  // Insert the stats into the state
  [STATS_UPDATE] : (previousState, action) => action.stats,
}

// ------------------------------------
// Reducer
// ------------------------------------
/**
 * Initial state for this store component
 */
export const initialState = {
  polls     : null,
  responses : null,
}

/**
 * The reducer for this store component
 *
 * @param  {object} state  The current state
 * @param  {object} action The action to perform on the state
 *
 * @return {object}        The modified state
 */
export default function statsReducer (state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type]

  return handler ? handler(state, action) : state
}
