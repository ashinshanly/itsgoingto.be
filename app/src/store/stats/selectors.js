import { prop, compose } from 'ramda'

/**
 * Returns a stat from the state
 *
 * @param  {object} state  App state
 * @param  {object} string The stat to select
 *
 * @return {boolean}       The stat
 */
export const statSelector = (state, stat) =>
  compose(
    prop(stat),
    prop('stats')
  )(state)
