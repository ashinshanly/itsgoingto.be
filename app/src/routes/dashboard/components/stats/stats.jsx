import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { isNil } from 'ramda'
import { fetchStats } from 'services/api'
import { statSelector } from 'store/stats/selectors'
import './stats.scss'

class Stats extends React.Component {
  componentDidMount = () => {
    const { fetchStats } = this.props

    fetchStats()
  }

  render() {
    const { polls, responses } = this.props

    return (
      <div className='info'>
        <div className='info-block-column'>
          <div className='info-block'>
            <div className='info-header'>
              Total Polls
            </div>
            <div className='info-body'>
              {!isNil(polls) ? polls : '-'}
            </div>
          </div>
        </div>
        <div className='info-block-column'>
          <div className='info-block'>
            <div className='info-header'>
              Total Responses
            </div>
            <div className='info-body'>
              {!isNil(responses) ? responses : '-'}
            </div>
          </div>
        </div>
      </div>
    )
  }
}

Stats.propTypes = {
  polls      : PropTypes.number,
  responses  : PropTypes.number,
  fetchStats : PropTypes.func.isRequired,
}

Stats.defaultProps = {
  polls     : undefined,
  responses : undefined,
}

const mapStateToProps = (state) => ({
  polls     : statSelector(state, 'polls'),
  responses : statSelector(state, 'responses'),
})

const mapDispatchToProps = (dispatch) => ({
  fetchStats : () => dispatch(fetchStats()),
})

export default connect(mapStateToProps, mapDispatchToProps)(Stats)