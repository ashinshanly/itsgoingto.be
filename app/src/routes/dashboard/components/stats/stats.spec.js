/* global expect, jest */
import React from 'react'
import { shallow } from 'enzyme'
import { Stats } from './stats'

const props = {
  fetchStats: jest.fn(),
}

describe('(Route) dashoard', () => {
  describe('(Component) stats', () => {
    let wrapper

    beforeEach(() => {
      wrapper = shallow(<Stats {...props} />)
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    describe('(Lifecycle) componentDidMount', () => {
      it('should call fetchStats', () => {
        expect(props.fetchStats).toHaveBeenCalledTimes(1)
      })
    })

    describe('(Render)', () => {
      describe('with no stats', () => {
        it('matches snapshot', () => {
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('with stats', () => {
        it('matches snapshot', () => {
          wrapper.setProps({ polls: 15, responses: 39 })
          expect(wrapper).toMatchSnapshot()
        })
      })
    })
  })
})
