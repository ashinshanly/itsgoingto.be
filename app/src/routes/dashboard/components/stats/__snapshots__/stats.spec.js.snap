// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) dashoard (Component) stats (Render) with no stats matches snapshot 1`] = `
<div
  className="info"
>
  <div
    className="info-block-column"
  >
    <div
      className="info-block"
    >
      <div
        className="info-header"
      >
        Total Polls
      </div>
      <div
        className="info-body"
      >
        -
      </div>
    </div>
  </div>
  <div
    className="info-block-column"
  >
    <div
      className="info-block"
    >
      <div
        className="info-header"
      >
        Total Responses
      </div>
      <div
        className="info-body"
      >
        -
      </div>
    </div>
  </div>
</div>
`;

exports[`(Route) dashoard (Component) stats (Render) with stats matches snapshot 1`] = `
<div
  className="info"
>
  <div
    className="info-block-column"
  >
    <div
      className="info-block"
    >
      <div
        className="info-header"
      >
        Total Polls
      </div>
      <div
        className="info-body"
      >
        15
      </div>
    </div>
  </div>
  <div
    className="info-block-column"
  >
    <div
      className="info-block"
    >
      <div
        className="info-header"
      >
        Total Responses
      </div>
      <div
        className="info-body"
      >
        39
      </div>
    </div>
  </div>
</div>
`;
